File: 백준/Gold/17143. 낚시왕/낚시왕.py
import sys
input = sys.stdin.readline

# 상어 잡기
def get():
    global site
    for i in range(R):
        for j in range(site, site+1):
            if graph[i][j]:
                x = graph[i][j][2]
                graph[i][j] = 0
                return x
    return 0

def shark_move():
    global graph

    graph2 = [[0] * C for _ in range(R)]
    for i in range(R):
        for j in range(C):
            if graph[i][j]!=0:
                # 상어가 있다면 상어의 위치를 옮기고, 새로운 graph에 저장
                nx, ny, ndir = next_dist(i,j,graph[i][j][0], graph[i][j][1])
                # 상어의 위치가 겹친다면
                if graph2[nx][ny]:
                        # 새로 온 상어와 기존에 있던 상어 중에 크기가 큰 상어를 남긴다.
                        graph2[nx][ny] = max(graph2[nx][ny], (graph[i][j][0], ndir, graph[i][j][2]), key=lambda x: x[2])
                else:
                    graph2[nx][ny] = graph[i][j][0], ndir, graph[i][j][2]

    graph = graph2

def next_dist(i,j,speed,dir):
    if dir == up or dir == down:
        cycle = (R - 1) * 2

        if dir == up:
            speed += 2 * (R - 1) - i
        else:
            speed += i

        speed %= cycle
        if speed >= R:
            return (cycle - speed, j, up)
        return(speed, j, down)

    else:
        cycle = (C - 1) * 2
        if dir == left:
            speed += 2 * (C - 1) - j
        else:
            speed += j

        speed %= cycle
        if speed >= C:
            return(i, cycle- speed, left)
        return(i, speed, right)


R,C,m = map(int, input().split())

graph = [[0] * C for _ in range(R)]

for _ in range(m):
   r,c,s,d,z = map(int, input().split()) 
   graph[r-1][c-1] = [s,d,z]

up, down, right, left = 1, 2, 3, 4

site = 0
ans = 0
for j in range(C):
    ans += get()
    shark_move()
    site += 1

print(ans)

File: 백준/Gold/17143. 낚시왕/낚시왕.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int r;
    static int c;
    static Shark[][] arr;
    static Shark[][] tmpArr;

    static int UP = 1;
    static int DOWN = 2;
    static int RIGHT = 3;
    static int LEFT = 4;

    static int site; // 낚시꾼의 위치
    static int ans; // 답

    static class Shark {
        int x;
        int y;
        int speed;
        int direction;
        int size;

        public Shark(int x, int y, int speed, int direction, int size) {
            this.x = x;
            this.y = y;
            this.speed = speed;
            this.direction = direction;
            this.size = size;
        }
    }

    static void setInputs() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        arr = new Shark[r][c];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            int speed = Integer.parseInt(st.nextToken());
            int direction = Integer.parseInt(st.nextToken());
            int size = Integer.parseInt(st.nextToken());
            // 상어 추가
            arr[x - 1][y - 1] = new Shark(x - 1, y - 1, speed, direction, size);
        }
    }

    static void catchShark() {
        for (int i = 0; i < r; i++) {
            if (arr[i][site] != null) {
                ans += arr[i][site].size;
                arr[i][site] = null;
                break;
            }
        }
        site++;
    }

    static void moveSharks() {
        tmpArr = new Shark[r][c];

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                Shark shark = arr[i][j];
                if (shark != null) {
                    moveIndividualShark(i, j, shark);

                    Shark originalShark = tmpArr[shark.x][shark.y];
                    // 만약 상어의 위치가 겹친다?
                    if (originalShark != null) {
                        // 새로운 상어의 사이즈가 크면 새로운 상어를 넣고 아니면 현행 유지
                        if (shark.size > originalShark.size) {
                            tmpArr[shark.x][shark.y] = shark;
                        } // null이 아니었는데 기존 상어가 더 크면 아무 것도 안해줘도 됨
                    } else {
                        // 해당 좌표에 아무 상어도 없었으면 새로운 상어 배치
                        tmpArr[shark.x][shark.y] = shark;
                    }
                }
            }
        }
        arr = tmpArr; // 배열 바꾸기
    }

    static void moveIndividualShark(int x, int y, Shark shark) {
        int dir = shark.direction;
        int speed = shark.speed;
        if (dir == UP || dir == DOWN) {
            int cycle = (r - 1) * 2;

            // Row가 0이면서 아래로 내려가는 형태로 시작을 고정
            if (dir == UP) {
                speed += cycle - x;
            } else {
                speed += x;
            }

            speed %= cycle; // 1회 반복 사이클은 계산하지 말 것
            if (speed >= r) {
                shark.x = cycle - speed; // 행 증가
                shark.direction = UP; // 반 사이클을 넘어가면 위로 올라감
            } else {
                shark.x = speed; // 행 증가
                shark.direction = DOWN;
            }
        } else if (dir == LEFT || dir == RIGHT) {
            int cycle = (c - 1) * 2;

            // Col이 0이면서 오른쪽으로 가는 형태로 시작을 고정
            if (dir == LEFT) {
                speed += cycle - y;
            } else {
                speed += y;
            }

            speed %= cycle;
            if (speed >= c) {
                shark.y = cycle - speed; // 열 증가
                shark.direction = LEFT; // 반 사이클을 넘으면 왼쪽으로 감
            } else {
                shark.y = speed; // 열 증가
                shark.direction = RIGHT;
            }
        }
    }

    public static void main(String[] args) throws Exception {
        setInputs();

        for (int i = 0; i < c; i++) {
            catchShark(); // 상어 잡기~
            moveSharks(); // 상어 도망가기~
        }

        System.out.println(ans);
    }
}